eip-0012.md

```diff
@@ -0,0 +1,277 @@
+
+### sign_data(addr: Address, message: String) -> Promise\<String> | DataSignError
+
+Signs the generic data {data} encoded as a hex string using the private key associated with {addr} if possible. If the wallet does not own this address, an error is returned. This should ask for user consent and produce an error if rejected. The wallet should also implement a message signing protocol such as the proposed [EmIP-005](https://github.com/Emurgo/EmIPs/blob/5b00fce84f31eb763892186eb9c88739ec809333/specs/emip-005.md) in order to make this endpoint safer for use, and make it harder for the user to accidentally sign data such as transactions, blocks, etc. Returns the signed data as a hex-encoded string.
+
+TBD: Hash algorithm/format to use.
+
+### submit_tx(tx: SignedTx) -> TxId | TxSendError
+
+Uses the wallet to send the given tx and returns the TxId for the dApp to track when/if it is included as its own responsibility. This is technically not mandatory but any wallet that is using such a bridge should have this functionality. Wallets can additionally choose to rate-limit this call, and can return an error if they did not or could not send the transaction.
+
+### add_external_box(box_id: BoxId)
+
+Informs the wallet of a box it should track, specifically for P2S/P2SH boxes that the wallet might not be aware of. The wallet does not need to follow this suggestion and can handle how it does this on its own.
+
+## Data Formats
+
+Most of these types are meant to be the same as the JSON types in the Ergo full node, which matches up as of now with the `sigma-rust` representations.
+
+### Address
+
+String as the standard base58 encoding of an address.
+
+### Box
+
+Same as `sigma-rust`:
+```
+{
+	boxId: BoxId,
+	value: Value,
+	ergoTree: ErgoTree,
+	assets: [TokenAmount],
+	additionalRegisters: { String => Constant },
+	creationHeight: int,
+	transactionId: TxId,
+	index: int,
+}
+```
+Additional registers have string ids "R4", "R5" to "R9".
+
+### BoxCandidate
+
+Same as `Box` without `transactionId` and `index` as those will be computed when the tx is completed.
+
+### BoxId
+
+Hex string of the box id.
+
+### Constant
+
+Uses `sigma-rust` rep - Hex-encoded bytes for `SigmaSerialize` of a constant.
+
+### ContextExtension
+
+Uses the full node JSON rep (id to hex-encoded binary buffer)
+```
+{
+	values: { String => String }
+}
+```
+or the empty `{}` object for P2PK inputs.
+
+### ErgoTree
+
+Uses `sigma-rust` rep - Hex string representing serialized bytes of the tree using `SigmaSerialize`
+
+### Input
+
+Read-only input (e.g. oracle usage or one yet to be signed) that is not spendable. Uses `sigma-rust` JSON rep:
```

Comment:
There's a difference between read-only input and ordinary input: while the former contains box id, the latter also may contain context extension.

Reply:
Updated to split into UnsignedInput/SignedInput/DataInput